server/game/cards/01_SOR/units/GuardianOfTheWhills.ts

```diff
@@ -0,0 +1,48 @@
+import AbilityHelper from '../../../AbilityHelper';
+import * as AbilityLimit from '../../../core/ability/AbilityLimit';
+import { Card } from '../../../core/card/Card';
+import { NonLeaderUnitCard } from '../../../core/card/NonLeaderUnitCard';
+import { StateWatcherRegistrar } from '../../../core/stateWatcher/StateWatcherRegistrar';
+import { CardsPlayedThisPhaseWatcher } from '../../../stateWatchers/CardsPlayedThisPhaseWatcher';
+
+export default class GuardianOfTheWhills extends NonLeaderUnitCard {
+    private cardsPlayedThisPhaseWatcher: CardsPlayedThisPhaseWatcher;
+
+    protected override getImplementationId() {
+        return {
+            id: '4166047484',
+            internalName: 'guardian-of-the-whills'
+        };
+    }
+
+    protected override setupStateWatchers(registrar: StateWatcherRegistrar): void {
+        this.cardsPlayedThisPhaseWatcher = AbilityHelper.stateWatchers.cardsPlayedThisPhase(registrar, this);
+    }
+
+    private isFirstUpgradePlayedOnThisCopy(card: Card, adjusterSource: Card): boolean {
+        if (!adjusterSource.isUnit()) {
+            return false;
+        }
+        const playedUpgradesOnThisCardThisPhase = this.cardsPlayedThisPhaseWatcher.getCardsPlayed((playedCardEntry) =>
+            playedCardEntry.card.isUpgrade() &&
+            playedCardEntry.card.parentCard === adjusterSource &&
+            playedCardEntry.card.parentCard.inPlayId === adjusterSource.inPlayId &&
+            playedCardEntry.card !== card
```

Comment:
We also need to check the play id of the upgrade card itself since it's possible that the same upgrade was returned to hand from play and played again. Test case for this would be good too - could defeat the attached unit and use Razor Crest to return the upgrade to hand

Reply:
Good spot, this will fail for sure since I filtered card === card so I'll skip the first time it was played